File: ios/SpentyAI/SpentyAI/Features/Invoices/InvoiceRepository.swift
import Foundation

struct InvoicePayload: Codable {
    let invoiceNumber: String?
    let customerId: String?
    let customerName: String?
    let invoiceDate: Date?
    let dueDate: Date?
    let lineItems: [InvoiceLineItemPayload]?
    let subtotal: Double?
    let taxAmount: Double?
    let grandTotal: Double?
    let notes: String?
    let termsConditions: String?
}

struct InvoiceLineItemPayload: Codable {
    let description: String?
    let hsnSac: String?
    let quantity: Double?
    let rate: Double?
    let taxPercent: Double?
    let amount: Double?
}

struct RecordPaymentPayload: Codable {
    let amount: Double
    let date: Date
    let paymentMethod: String?
    let accountId: String?
    let note: String?
}

private struct InvoiceListResponse: Codable {
    let invoices: [Invoice]
    let total: Int?

    enum CodingKeys: String, CodingKey {
        case invoices
        case items
        case total
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        // Backend may send "invoices" or "items" key
        if let invoices = try? container.decode([Invoice].self, forKey: .invoices) {
            self.invoices = invoices
        } else if let items = try? container.decode([Invoice].self, forKey: .items) {
            self.invoices = items
        } else {
            self.invoices = []
        }
        self.total = try? container.decodeIfPresent(Int.self, forKey: .total)
    }
}

struct InvoiceRepository {

    private let api = APIClient.shared

    // MARK: - CRUD

    func fetchAll() async throws -> [Invoice] {
        let response: InvoiceListResponse = try await api.get(APIEndpoints.invoices)
        return response.invoices
    }

    func fetch(id: String) async throws -> Invoice {
        try await api.get(APIEndpoints.invoice(id))
    }

    func create(_ payload: InvoicePayload) async throws -> Invoice {
        try await api.post(APIEndpoints.invoices, body: payload)
    }

    func update(id: String, _ payload: InvoicePayload) async throws -> Invoice {
        try await api.put(APIEndpoints.invoice(id), body: payload)
    }

    func delete(id: String) async throws {
        let _: MessageResponse = try await api.delete(APIEndpoints.invoice(id))
    }

    // MARK: - Actions

    func recordPayment(id: String, _ payload: RecordPaymentPayload) async throws -> Invoice {
        try await api.post(APIEndpoints.invoiceRecordPayment(id), body: payload)
    }

    func markPaid(id: String) async throws -> Invoice {
        try await api.post(APIEndpoints.invoiceMarkPaid(id))
    }

    func duplicate(id: String) async throws -> Invoice {
        try await api.post(APIEndpoints.invoiceDuplicate(id))
    }

    // MARK: - Stats & Aggregates

    func fetchNextNumber() async throws -> InvoiceNextNumber {
        try await api.get(APIEndpoints.invoicesNextNumber)
    }

    func fetchStats() async throws -> InvoiceStats {
        try await api.get(APIEndpoints.invoicesStats)
    }

    func fetchCount() async throws -> InvoiceCountResponse {
        try await api.get(APIEndpoints.invoicesCount)
    }

    func fetchDebtors() async throws -> [InvoiceDebtor] {
        try await api.get(APIEndpoints.invoicesDebtors)
    }

    func fetchAging() async throws -> [InvoiceAgingBucket] {
        try await api.get(APIEndpoints.invoicesAging)
    }

    func fetchSalesByCustomer() async throws -> [InvoiceSalesByCustomer] {
        try await api.get(APIEndpoints.invoicesSalesByCustomer)
    }

    // MARK: - PDF

    func fetchPDF(id: String) async throws -> Data {
        try await api.getRaw(APIEndpoints.invoicePDF(id))
    }

    // MARK: - Customers & Accounts (for pickers)

    func fetchCustomers() async throws -> [Customer] {
        let response: CustomerListResponse = try await api.get(APIEndpoints.customers)
        return response.items
    }

    func fetchAccounts() async throws -> [Account] {
        try await api.get(APIEndpoints.accounts)
    }
}

File: android-native/app/src/main/java/com/spentyai/app/features/invoices/InvoiceRepository.kt
package com.spentyai.app.features.invoices

import com.spentyai.app.core.models.Account
import com.spentyai.app.core.models.Customer
import com.spentyai.app.core.models.Invoice
import com.spentyai.app.core.network.ApiClient
import com.spentyai.app.core.network.ApiResult
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray

// ---------- Response / Payload models ----------

@Serializable
data class InvoiceStats(
    @SerialName("total_invoiced") val totalInvoiced: Double = 0.0,
    @SerialName("total_paid") val totalPaid: Double = 0.0,
    @SerialName("total_outstanding") val totalOutstanding: Double = 0.0,
    @SerialName("total_overdue") val totalOverdue: Double = 0.0
)

@Serializable
data class InvoiceDebtor(
    @SerialName("customer_id") val customerId: String? = null,
    @SerialName("customer_name") val customerName: String? = null,
    @SerialName("total_outstanding") val totalOutstanding: Double = 0.0,
    @SerialName("invoice_count") val invoiceCount: Int = 0
)

@Serializable
data class InvoiceAgingBucket(
    val label: String? = null,
    val amount: Double = 0.0,
    val count: Int = 0
)

@Serializable
data class InvoiceSalesByCustomer(
    @SerialName("customer_id") val customerId: String? = null,
    @SerialName("customer_name") val customerName: String? = null,
    @SerialName("total_sales") val totalSales: Double = 0.0,
    @SerialName("invoice_count") val invoiceCount: Int = 0
)

@Serializable
data class InvoiceNextNumber(
    @SerialName("next_number") val nextNumber: String? = null
)

class InvoiceRepository(private val apiClient: ApiClient) {

    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
        isLenient = true
        encodeDefaults = true
    }

    // CRUD

    suspend fun fetchAll(): ApiResult<List<Invoice>> =
        apiClient.safeApiCall { apiClient.endpoints.getInvoices() }

    suspend fun fetch(id: String): ApiResult<Invoice> =
        apiClient.safeApiCall { apiClient.endpoints.getInvoice(id) }

    suspend fun create(payload: JsonObject): ApiResult<Invoice> =
        apiClient.safeApiCall { apiClient.endpoints.createInvoice(payload) }

    suspend fun update(id: String, payload: JsonObject): ApiResult<Invoice> =
        apiClient.safeApiCall { apiClient.endpoints.updateInvoice(id, payload) }

    suspend fun delete(id: String): ApiResult<JsonObject> =
        apiClient.safeApiCall { apiClient.endpoints.deleteInvoice(id) }

    // Actions

    suspend fun recordPayment(id: String, payload: JsonObject): ApiResult<Invoice> =
        apiClient.safeApiCall { apiClient.endpoints.markInvoicePaid(id, payload) }

    suspend fun markPaid(id: String): ApiResult<Invoice> =
        apiClient.safeApiCall {
            apiClient.endpoints.markInvoicePaid(id, buildJsonObject { put("mark_fully_paid", true) })
        }

    // Stats & Aggregates (these use the base invoices endpoint with query params, or dedicated endpoints)
    // For now, we compute stats client-side from the invoice list since the Android API doesn't have
    // dedicated stats endpoints yet. The iOS app uses custom endpoints.

    suspend fun fetchCustomers(): ApiResult<List<Customer>> =
        apiClient.safeApiCall { apiClient.endpoints.getCustomers() }

    suspend fun fetchAccounts(): ApiResult<List<Account>> =
        apiClient.safeApiCall { apiClient.endpoints.getAccounts() }

    // Build payload

    fun buildInvoicePayload(
        invoiceNumber: String,
        customerId: String?,
        customerName: String?,
        issueDate: String,
        dueDate: String,
        lineItems: List<InvoiceFormLineItem>,
        subtotal: Double,
        taxAmount: Double,
        totalCgst: Double,
        totalSgst: Double,
        totalIgst: Double,
        grandTotal: Double,
        notes: String?,
        terms: String?
    ): JsonObject = buildJsonObject {
        put("invoice_number", invoiceNumber)
        customerId?.let { put("customer_id", it) }
        customerName?.let { put("customer_name", it) }
        put("issue_date", issueDate)
        put("due_date", dueDate)
        put("subtotal", subtotal)
        put("tax_amount", taxAmount)
        put("total_cgst", totalCgst)
        put("total_sgst", totalSgst)
        put("total_igst", totalIgst)
        put("total", grandTotal)
        notes?.let { put("notes", it) }
        terms?.let { put("terms", it) }
        putJsonArray("line_items") {
            lineItems.forEach { item ->
                add(buildJsonObject {
                    put("description", item.description)
                    if (item.hsnSac.isNotBlank()) put("hsn_sac", item.hsnSac)
                    put("quantity", item.quantity)
                    put("unit_price", item.rate)
                    put("tax_rate", item.taxPercent)
                    put("amount", item.taxableAmount)
                })
            }
        }
    }

    fun buildRecordPaymentPayload(
        amount: Double,
        date: String,
        method: String,
        accountId: String?,
        note: String?
    ): JsonObject = buildJsonObject {
        put("amount", amount)
        put("date", date)
        put("payment_method", method)
        accountId?.let { put("account_id", it) }
        note?.let { put("note", it) }
    }
}

// Form line item used locally in the form
data class InvoiceFormLineItem(
    val id: String = java.util.UUID.randomUUID().toString(),
    var description: String = "",
    var hsnSac: String = "",
    var quantity: Double = 1.0,
    var rate: Double = 0.0,
    var taxPercent: Double = 18.0
) {
    val taxableAmount: Double get() = quantity * rate
    val taxAmount: Double get() = taxableAmount * taxPercent / 100.0
    val lineTotal: Double get() = taxableAmount + taxAmount
}
